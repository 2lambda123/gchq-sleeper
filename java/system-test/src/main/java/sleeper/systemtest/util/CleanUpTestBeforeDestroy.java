/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.util;

import sleeper.configuration.properties.local.LoadLocalProperties;
import sleeper.configuration.properties.table.TableProperties;
import sleeper.status.update.CleanUpBeforeDestroy;
import sleeper.systemtest.SystemTestProperties;

import java.nio.file.Path;
import java.util.List;
import java.util.stream.Collectors;

import static sleeper.systemtest.SystemTestProperty.SYSTEM_TEST_CLUSTER_NAME;

public class CleanUpTestBeforeDestroy {

    private CleanUpTestBeforeDestroy() {
    }

    public static void main(String[] args) {
        if (1 != args.length) {
            throw new IllegalArgumentException("Usage: <configuration directory>");
        }
        Path baseDir = Path.of(args[0]);
        SystemTestProperties instanceProperties = LoadLocalProperties.loadInstancePropertiesFromDirectory(new SystemTestProperties(), baseDir);
        List<TableProperties> tablePropertiesList = LoadLocalProperties
                .loadTablesFromDirectory(instanceProperties, baseDir).collect(Collectors.toList());

        CleanUpBeforeDestroy.cleanUp(instanceProperties, tablePropertiesList,
                List.of(instanceProperties.get(SYSTEM_TEST_CLUSTER_NAME)));
    }
}
