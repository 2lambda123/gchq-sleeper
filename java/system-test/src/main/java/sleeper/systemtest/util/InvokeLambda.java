/*
 * Copyright 2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.util;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import sleeper.configuration.properties.InstanceProperty;
import sleeper.systemtest.SystemTestProperties;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.services.lambda.LambdaClient;

import java.io.IOException;

public class InvokeLambda {

    private InvokeLambda() {
    }

    public static void forInstance(String instanceId, InstanceProperty lambdaFunctionProperty) throws IOException {

        AmazonS3 s3Client = AmazonS3ClientBuilder.defaultClient();
        SystemTestProperties systemTestProperties = new SystemTestProperties();
        systemTestProperties.loadFromS3GivenInstanceId(s3Client, instanceId);
        s3Client.shutdown();

        try (LambdaClient lambda = LambdaClient.create()) {
            lambda.invoke(builder -> builder
                    .functionName(systemTestProperties.get(lambdaFunctionProperty))
                    .payload(SdkBytes.fromUtf8String("{}")));
        }
    }
}
