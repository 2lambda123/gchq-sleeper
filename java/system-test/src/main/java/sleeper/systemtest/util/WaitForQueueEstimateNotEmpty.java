/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.util;

import com.amazonaws.services.sqs.AmazonSQS;
import com.amazonaws.services.sqs.model.GetQueueAttributesRequest;
import com.amazonaws.services.sqs.model.GetQueueAttributesResult;
import com.amazonaws.services.sqs.model.QueueAttributeName;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.configuration.properties.InstanceProperties;
import sleeper.configuration.properties.InstanceProperty;

public class WaitForQueueEstimateNotEmpty {
    private static final Logger LOGGER = LoggerFactory.getLogger(WaitForQueueEstimateNotEmpty.class);
    private static final long POLL_INTERVAL_MILLIS = 5000;
    private static final int MAX_POLLS = 12;

    private final AmazonSQS sqsClient;
    private final String queueUrl;
    private final PollWithRetries poll = PollWithRetries.intervalAndMaxPolls(POLL_INTERVAL_MILLIS, MAX_POLLS);

    public WaitForQueueEstimateNotEmpty(AmazonSQS sqsClient, InstanceProperties instanceProperties, InstanceProperty queueProperty) {
        this.sqsClient = sqsClient;
        this.queueUrl = instanceProperties.get(queueProperty);
    }

    public void pollUntilFinished() throws InterruptedException {
        LOGGER.info("Waiting for messages on {}", queueUrl);
        poll.pollUntil("queue estimated not empty", this::isEstimateNotEmpty);
    }

    private boolean isEstimateNotEmpty() {
        int approximateMessages = getApproximateMessages();
        LOGGER.info("Approximate number of messages: {}", approximateMessages);
        return approximateMessages > 0;
    }

    private int getApproximateMessages() {
        GetQueueAttributesResult result = sqsClient.getQueueAttributes(new GetQueueAttributesRequest()
                .withQueueUrl(queueUrl)
                .withAttributeNames(QueueAttributeName.ApproximateNumberOfMessages));
        return getInt(result, QueueAttributeName.ApproximateNumberOfMessages);
    }

    private static int getInt(GetQueueAttributesResult result, QueueAttributeName attribute) {
        return Integer.parseInt(result.getAttributes().get(attribute.toString()));
    }
}
