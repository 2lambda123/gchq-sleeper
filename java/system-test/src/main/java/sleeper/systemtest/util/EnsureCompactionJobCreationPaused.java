/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.util;

import com.amazonaws.services.cloudwatchevents.AmazonCloudWatchEvents;
import com.amazonaws.services.cloudwatchevents.AmazonCloudWatchEventsClientBuilder;
import com.amazonaws.services.cloudwatchevents.model.DescribeRuleRequest;
import com.amazonaws.services.cloudwatchevents.model.DescribeRuleResult;
import com.amazonaws.services.cloudwatchevents.model.RuleState;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;

import sleeper.configuration.properties.InstanceProperties;
import sleeper.util.ClientUtils;

import java.io.IOException;

import static sleeper.configuration.properties.SystemDefinedInstanceProperty.COMPACTION_JOB_CREATION_CLOUDWATCH_RULE;

public class EnsureCompactionJobCreationPaused {

    private EnsureCompactionJobCreationPaused() {
    }

    public static void main(String[] args) throws IOException {
        if (args.length != 1) {
            System.out.println("Usage: <instance id>");
            return;
        }

        AmazonS3 amazonS3 = AmazonS3ClientBuilder.defaultClient();
        InstanceProperties instanceProperties = ClientUtils.getInstanceProperties(amazonS3, args[0]);
        amazonS3.shutdown();

        AmazonCloudWatchEvents cwClient = AmazonCloudWatchEventsClientBuilder.defaultClient();

        String ruleName = instanceProperties.get(COMPACTION_JOB_CREATION_CLOUDWATCH_RULE);

        DescribeRuleResult result = cwClient.describeRule(new DescribeRuleRequest().withName(ruleName));
        RuleState state = RuleState.fromValue(result.getState());
        if (state != RuleState.DISABLED) {
            System.err.println("Compaction job creation CloudWatch rule is not paused");
            System.exit(1);
        }
    }
}
