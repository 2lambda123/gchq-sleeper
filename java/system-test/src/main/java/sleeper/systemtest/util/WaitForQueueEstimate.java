/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.util;

import com.amazonaws.services.sqs.AmazonSQS;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.clients.util.PollWithRetries;
import sleeper.compaction.job.CompactionJobStatusStore;
import sleeper.configuration.properties.InstanceProperties;
import sleeper.configuration.properties.InstanceProperty;
import sleeper.job.common.QueueMessageCount;

import java.util.function.Predicate;

import static sleeper.job.common.QueueMessageCount.withSqsClient;

public class WaitForQueueEstimate {
    private static final Logger LOGGER = LoggerFactory.getLogger(WaitForQueueEstimate.class);

    private final QueueMessageCount.Client queueClient;
    private final Predicate<QueueMessageCount> isFinished;
    private final String description;
    private final String queueUrl;

    public static WaitForQueueEstimate notEmpty(
            QueueMessageCount.Client queueClient, InstanceProperties instanceProperties, InstanceProperty queueProperty) {
        String queueUrl = instanceProperties.get(queueProperty);
        return new WaitForQueueEstimate(queueClient, queueUrl,
                estimate -> estimate.getApproximateNumberOfMessages() > 0,
                "estimate not empty for queue " + queueUrl);
    }

    public static WaitForQueueEstimate isEmpty(
            AmazonSQS sqsClient, InstanceProperties instanceProperties, InstanceProperty queueProperty) {
        String queueUrl = instanceProperties.get(queueProperty);
        return new WaitForQueueEstimate(withSqsClient(sqsClient), queueUrl,
                estimate -> estimate.getApproximateNumberOfMessages() == 0,
                "estimate empty for queue " + queueUrl);
    }

    public static WaitForQueueEstimate containsUnfinishedJobs(
            AmazonSQS sqsClient, InstanceProperties instanceProperties, InstanceProperty queueProperty,
            CompactionJobStatusStore statusStore, String tableName) {
        int unfinished = statusStore.getUnfinishedJobs(tableName).size();
        LOGGER.info("Found {} unfinished compaction jobs", unfinished);
        return new WaitForQueueEstimate(withSqsClient(sqsClient), instanceProperties, queueProperty,
                estimate -> estimate.getApproximateNumberOfMessages() >= unfinished,
                "queue estimate matching unfinished compaction jobs");
    }

    private WaitForQueueEstimate(QueueMessageCount.Client queueClient,
                                 InstanceProperties properties, InstanceProperty queueProperty,
                                 Predicate<QueueMessageCount> isFinished, String description) {
        this(queueClient, properties.get(queueProperty), isFinished, description);
    }

    private WaitForQueueEstimate(QueueMessageCount.Client queueClient, String queueUrl,
                                 Predicate<QueueMessageCount> isFinished, String description) {
        this.queueClient = queueClient;
        this.queueUrl = queueUrl;
        this.isFinished = isFinished;
        this.description = description;
    }

    public void pollUntilFinished(PollWithRetries poll) throws InterruptedException {
        LOGGER.info("Waiting until {}", description);
        poll.pollUntil(description, this::isFinished);
    }

    private boolean isFinished() {
        QueueMessageCount count = queueClient.getQueueMessageCount(queueUrl);
        LOGGER.info("Message count: {}", count);
        return isFinished.test(count);
    }
}
