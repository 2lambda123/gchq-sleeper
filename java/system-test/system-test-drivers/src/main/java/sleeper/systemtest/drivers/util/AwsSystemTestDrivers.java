/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.drivers.util;

import sleeper.systemtest.drivers.compaction.AwsCompactionDriver;
import sleeper.systemtest.drivers.compaction.AwsCompactionReportsDriver;
import sleeper.systemtest.drivers.ingest.AwsDataGenerationTasksDriver;
import sleeper.systemtest.drivers.ingest.AwsDirectIngestDriver;
import sleeper.systemtest.drivers.ingest.AwsIngestBatcherDriver;
import sleeper.systemtest.drivers.ingest.AwsIngestByQueueDriver;
import sleeper.systemtest.drivers.ingest.AwsIngestReportsDriver;
import sleeper.systemtest.drivers.ingest.AwsInvokeIngestTasksDriver;
import sleeper.systemtest.drivers.ingest.AwsPurgeQueueDriver;
import sleeper.systemtest.drivers.ingest.DirectEmrServerlessDriver;
import sleeper.systemtest.drivers.instance.AwsSleeperInstanceDriver;
import sleeper.systemtest.drivers.instance.AwsSleeperInstanceTablesDriver;
import sleeper.systemtest.drivers.instance.AwsSystemTestDeploymentDriver;
import sleeper.systemtest.drivers.metrics.AwsTableMetricsDriver;
import sleeper.systemtest.drivers.partitioning.AwsPartitionReportDriver;
import sleeper.systemtest.drivers.partitioning.AwsPartitionSplittingDriver;
import sleeper.systemtest.drivers.python.PythonBulkImportDriver;
import sleeper.systemtest.drivers.python.PythonIngestDriver;
import sleeper.systemtest.drivers.python.PythonIngestLocalFileDriver;
import sleeper.systemtest.drivers.python.PythonQueryDriver;
import sleeper.systemtest.drivers.query.DirectQueryDriver;
import sleeper.systemtest.drivers.query.S3ResultsDriver;
import sleeper.systemtest.drivers.query.SQSQueryDriver;
import sleeper.systemtest.drivers.sourcedata.AwsGeneratedIngestSourceFilesDriver;
import sleeper.systemtest.drivers.sourcedata.AwsIngestSourceFilesDriver;
import sleeper.systemtest.dsl.compaction.SystemTestCompaction;
import sleeper.systemtest.dsl.ingest.IngestByQueue;
import sleeper.systemtest.dsl.ingest.SystemTestIngest;
import sleeper.systemtest.dsl.instance.DeployedSleeperInstances;
import sleeper.systemtest.dsl.instance.SleeperInstanceDriver;
import sleeper.systemtest.dsl.instance.SleeperInstanceTablesDriver;
import sleeper.systemtest.dsl.instance.SystemTestDeploymentContext;
import sleeper.systemtest.dsl.instance.SystemTestInstanceContext;
import sleeper.systemtest.dsl.instance.SystemTestParameters;
import sleeper.systemtest.dsl.metrics.SystemTestMetrics;
import sleeper.systemtest.dsl.partitioning.SystemTestPartitioning;
import sleeper.systemtest.dsl.python.SystemTestPythonApi;
import sleeper.systemtest.dsl.query.SystemTestQuery;
import sleeper.systemtest.dsl.reporting.ReportingContext;
import sleeper.systemtest.dsl.reporting.SystemTestReporting;
import sleeper.systemtest.dsl.reporting.SystemTestReports;
import sleeper.systemtest.dsl.sourcedata.GeneratedIngestSourceFilesDriver;
import sleeper.systemtest.dsl.sourcedata.IngestSourceFilesContext;
import sleeper.systemtest.dsl.sourcedata.SystemTestCluster;
import sleeper.systemtest.dsl.sourcedata.SystemTestSourceFiles;
import sleeper.systemtest.dsl.util.PurgeQueueDriver;
import sleeper.systemtest.dsl.util.SystemTestDrivers;

import java.nio.file.Path;

public class AwsSystemTestDrivers implements SystemTestDrivers {
    private final SystemTestClients clients = new SystemTestClients();
    private final SystemTestParameters parameters;
    private final SystemTestDeploymentContext systemTestContext;
    private final SystemTestInstanceContext instanceContext;
    private final IngestSourceFilesContext sourceFilesContext;
    private final ReportingContext reportingContext;

    public AwsSystemTestDrivers(SystemTestParameters parameters) {
        this.parameters = parameters;
        systemTestContext = new SystemTestDeploymentContext(
                parameters, new AwsSystemTestDeploymentDriver(parameters, clients));
        SleeperInstanceDriver instanceDriver = new AwsSleeperInstanceDriver(parameters, clients);
        SleeperInstanceTablesDriver tablesDriver = new AwsSleeperInstanceTablesDriver(clients);
        DeployedSleeperInstances deployedInstances = new DeployedSleeperInstances(
                parameters, systemTestContext, instanceDriver, tablesDriver);
        instanceContext = new SystemTestInstanceContext(parameters, deployedInstances, instanceDriver, tablesDriver);
        sourceFilesContext = new IngestSourceFilesContext(systemTestContext, instanceContext);
        reportingContext = new ReportingContext(parameters);
    }

    @Override
    public SystemTestDeploymentContext getSystemTestContext() {
        return systemTestContext;
    }

    @Override
    public SystemTestInstanceContext getInstanceContext() {
        return instanceContext;
    }

    @Override
    public IngestSourceFilesContext getSourceFilesContext() {
        return sourceFilesContext;
    }

    @Override
    public ReportingContext getReportingContext() {
        return reportingContext;
    }

    @Override
    public GeneratedIngestSourceFilesDriver generatedSourceFilesDriver() {
        return new AwsGeneratedIngestSourceFilesDriver(systemTestContext, clients.getS3V2());
    }

    @Override
    public SystemTestSourceFiles sourceFiles() {
        return new SystemTestSourceFiles(instanceContext, sourceFilesContext,
                new AwsIngestSourceFilesDriver(sourceFilesContext));
    }

    @Override
    public SystemTestPartitioning partitioning() {
        return new SystemTestPartitioning(instanceContext,
                new AwsPartitionSplittingDriver(instanceContext, clients.getLambda()));
    }

    @Override
    public SystemTestIngest ingest() {
        return new SystemTestIngest(instanceContext, sourceFilesContext,
                new AwsDirectIngestDriver(instanceContext),
                new IngestByQueue(instanceContext, new AwsIngestByQueueDriver(clients)),
                new DirectEmrServerlessDriver(instanceContext, clients),
                new AwsIngestBatcherDriver(instanceContext, sourceFilesContext, clients),
                new AwsInvokeIngestTasksDriver(instanceContext, clients),
                AwsWaitForJobs.forIngest(instanceContext, clients.getDynamoDB()),
                AwsWaitForJobs.forBulkImport(instanceContext, clients.getDynamoDB()));
    }

    @Override
    public SystemTestQuery query() {
        return new SystemTestQuery(instanceContext,
                SQSQueryDriver.allTablesDriver(instanceContext, clients),
                DirectQueryDriver.allTablesDriver(instanceContext),
                new S3ResultsDriver(instanceContext, clients.getS3()));
    }

    @Override
    public SystemTestCompaction compaction() {
        return new SystemTestCompaction(
                new AwsCompactionDriver(instanceContext, clients),
                AwsWaitForJobs.forCompaction(instanceContext, clients.getDynamoDB()));
    }

    @Override
    public SystemTestReporting reporting() {
        return new SystemTestReporting(reportingContext,
                new AwsIngestReportsDriver(instanceContext, clients),
                new AwsCompactionReportsDriver(instanceContext, clients.getDynamoDB()));
    }

    @Override
    public SystemTestMetrics metrics() {
        return new SystemTestMetrics(new AwsTableMetricsDriver(instanceContext, reportingContext, clients));
    }

    @Override
    public SystemTestReports.SystemTestBuilder reportsForExtension() {
        return SystemTestReports.builder(reportingContext,
                new AwsPartitionReportDriver(instanceContext),
                new AwsIngestReportsDriver(instanceContext, clients),
                new AwsCompactionReportsDriver(instanceContext, clients.getDynamoDB()));
    }

    @Override
    public SystemTestCluster systemTestCluster() {
        return new SystemTestCluster(systemTestContext,
                new AwsDataGenerationTasksDriver(systemTestContext, instanceContext, clients.getEcs()),
                new IngestByQueue(instanceContext, new AwsIngestByQueueDriver(clients)),
                new AwsGeneratedIngestSourceFilesDriver(systemTestContext, clients.getS3V2()),
                new AwsInvokeIngestTasksDriver(instanceContext, clients),
                AwsWaitForJobs.forIngest(instanceContext, clients.getDynamoDB()),
                AwsWaitForJobs.forBulkImport(instanceContext, clients.getDynamoDB()));
    }

    @Override
    public SystemTestPythonApi pythonApi() {
        Path pythonDir = parameters.getPythonDirectory();
        return new SystemTestPythonApi(instanceContext,
                new PythonIngestDriver(instanceContext, pythonDir),
                new PythonIngestLocalFileDriver(instanceContext, pythonDir),
                new PythonBulkImportDriver(instanceContext, pythonDir),
                new AwsInvokeIngestTasksDriver(instanceContext, clients),
                AwsWaitForJobs.forIngest(instanceContext, clients.getDynamoDB()),
                AwsWaitForJobs.forBulkImport(instanceContext, clients.getDynamoDB()),
                new PythonQueryDriver(instanceContext, pythonDir));
    }

    @Override
    public PurgeQueueDriver purgeQueueDriver() {
        return new AwsPurgeQueueDriver(instanceContext, clients.getSqs());
    }
}
