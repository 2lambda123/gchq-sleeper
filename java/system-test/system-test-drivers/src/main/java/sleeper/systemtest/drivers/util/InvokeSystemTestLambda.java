/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.drivers.util;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import software.amazon.awssdk.http.apache.ApacheHttpClient;
import software.amazon.awssdk.services.lambda.LambdaClient;

import sleeper.clients.deploy.InvokeLambda;
import sleeper.configuration.properties.InstanceProperties;
import sleeper.configuration.properties.InstanceProperty;
import sleeper.systemtest.configuration.SystemTestProperties;

import java.io.IOException;
import java.time.Duration;

public class InvokeSystemTestLambda {

    private InvokeSystemTestLambda() {
    }

    public static void forInstance(String instanceId, InstanceProperty lambdaFunctionProperty) throws IOException {
        try (LambdaClient lambdaClient = createSystemTestLambdaClient()) {
            AmazonS3 s3Client = AmazonS3ClientBuilder.defaultClient();
            SystemTestProperties systemTestProperties = new SystemTestProperties();
            systemTestProperties.loadFromS3GivenInstanceId(s3Client, instanceId);
            s3Client.shutdown();
            client(lambdaClient, systemTestProperties).invokeLambda(lambdaFunctionProperty);
        }
    }

    public static Client client(LambdaClient lambdaClient, InstanceProperties instanceProperties) {
        return lambdaFunctionProperty ->
                InvokeLambda.invokeWith(lambdaClient, instanceProperties.get(lambdaFunctionProperty));
    }

    public static LambdaClient createSystemTestLambdaClient() {
        return LambdaClient.builder()
                .overrideConfiguration(builder -> builder
                        .apiCallTimeout(Duration.ofMinutes(5))
                        .apiCallAttemptTimeout(Duration.ofMinutes(5)))
                .httpClientBuilder(ApacheHttpClient.builder()
                        .socketTimeout(Duration.ofMinutes(5)))
                .build();
    }

    public interface Client {
        void invokeLambda(InstanceProperty lambdaFunctionProperty);
    }
}
