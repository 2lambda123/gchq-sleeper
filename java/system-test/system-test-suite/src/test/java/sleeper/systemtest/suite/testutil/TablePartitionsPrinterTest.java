/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.suite.testutil;

import org.junit.jupiter.api.Test;

import sleeper.core.partition.PartitionsBuilder;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.LongType;

import java.util.List;
import java.util.Map;

import static org.approvaltests.Approvals.verify;
import static org.assertj.core.api.Assertions.assertThat;
import static sleeper.core.schema.SchemaTestHelper.schemaWithKey;

public class TablePartitionsPrinterTest {

    private final Schema schema = schemaWithKey("key", new LongType());
    private final PartitionsBuilder partitions = new PartitionsBuilder(schema);

    @Test
    void shouldPrintTreeWith8LevelsOfSplits() {
        partitions.rootFirst("root")
                .splitToNewChildren("root", "L", "R", 50L)
                .splitToNewChildren("L", "LL", "LR", 25L)
                .splitToNewChildren("R", "RL", "RR", 75L)
                .splitToNewChildren("LL", "LLL", "LLR", 12L)
                .splitToNewChildren("LR", "LRL", "LRR", 37L)
                .splitToNewChildren("RL", "RLL", "RLR", 62L)
                .splitToNewChildren("RR", "RRL", "RRR", 87L);

        verify(TablePartitionsPrinter.printPartitions(schema, partitions.buildTree()));
    }

    @Test
    void shouldRenamePartitionsByLocation() {
        partitions.rootFirst("base")
                .splitToNewChildren("base", "l", "r", 50L)
                .splitToNewChildren("l", "ll", "lr", 25L)
                .splitToNewChildren("r", "rl", "rr", 75L)
                .splitToNewChildren("ll", "1", "2", 12L)
                .splitToNewChildren("lr", "3", "4", 37L)
                .splitToNewChildren("rl", "5", "6", 62L)
                .splitToNewChildren("rr", "7", "8", 87L);

        verify(TablePartitionsPrinter.printPartitions(schema, partitions.buildTree()));
    }

    @Test
    void shouldPrintDifferentPartitionsForOneTable() {
        PartitionsBuilder partitions1 = new PartitionsBuilder(schema).rootFirst("A")
                .splitToNewChildren("A", "B", "C", 10L);
        PartitionsBuilder partitions2 = new PartitionsBuilder(schema).rootFirst("1")
                .splitToNewChildren("1", "2", "3", 20L);

        verify(TablePartitionsPrinter.printTablePartitionsExpectingIdentical(schema, Map.of(
                "table-1", partitions1.buildTree(),
                "table-2", partitions2.buildTree(),
                "table-3", partitions1.buildTree())));
    }

    @Test
    void shouldPrintExpectedForTables() {
        partitions.rootFirst("A")
                .splitToNewChildren("A", "B", "C", 10L);

        assertThat(TablePartitionsPrinter.printExpectedPartitionsForAllTables(schema, List.of("table-1", "table-2"), partitions.buildTree()))
                .isEqualTo(TablePartitionsPrinter.printTablePartitionsExpectingIdentical(schema, Map.of(
                        "table-1", partitions.buildTree(), "table-2", partitions.buildTree())));
    }
}
