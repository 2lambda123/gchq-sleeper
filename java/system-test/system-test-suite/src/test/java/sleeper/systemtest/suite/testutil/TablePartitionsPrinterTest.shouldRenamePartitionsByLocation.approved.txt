Partition at LLL with ID: 1
{"key":{"min":-9223372036854775808,"minInclusive":true,"max":12,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at LLR with ID: 2
{"key":{"min":12,"minInclusive":true,"max":25,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at LRL with ID: 3
{"key":{"min":25,"minInclusive":true,"max":37,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at LRR with ID: 4
{"key":{"min":37,"minInclusive":true,"max":50,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at RLL with ID: 5
{"key":{"min":50,"minInclusive":true,"max":62,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at RLR with ID: 6
{"key":{"min":62,"minInclusive":true,"max":75,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at RRL with ID: 7
{"key":{"min":75,"minInclusive":true,"max":87,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at RRR with ID: 8
{"key":{"min":87,"minInclusive":true,"max":null,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at LL with ID: ll
{"key":{"min":-9223372036854775808,"minInclusive":true,"max":25,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at LR with ID: lr
{"key":{"min":25,"minInclusive":true,"max":50,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at RL with ID: rl
{"key":{"min":50,"minInclusive":true,"max":75,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at RR with ID: rr
{"key":{"min":75,"minInclusive":true,"max":null,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at L with ID: l
{"key":{"min":-9223372036854775808,"minInclusive":true,"max":50,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at R with ID: r
{"key":{"min":50,"minInclusive":true,"max":null,"maxInclusive":false},"stringsBase64Encoded":true}
Partition at root with ID: base
{"key":{"min":-9223372036854775808,"minInclusive":true,"max":null,"maxInclusive":false},"stringsBase64Encoded":true}
